runtime/near-vm-runner/src/wasmer2_runner.rs

```diff
@@ -378,29 +332,45 @@ impl Wasmer2VM {
                             .with_counter(gas)
                             .with_stack_limit(self.config.limit_config.wasmer2_stack_limit),
                     )
-                    .map_err(|err| translate_instantiation_error(err, import.vmlogic))?;
+                    .map_err(|err| {
+                        use wasmer_engine::InstantiationError::{self, *};
+                        match err.downcast_ref::<InstantiationError>() {
+                            Some(Start(err)) => translate_runtime_error(err.clone(), import.vmlogic),
+                            Some(Link(e)) => VMError::FunctionCallError(FunctionCallError::LinkError { msg: e.to_string() }),
+                            Some(CreateInstance(e)) => VMError::FunctionCallError(FunctionCallError::LinkError { msg: e.to_string() }),
+                            Some(CpuFeature(e)) => {
+                                panic!("host does not support the CPU features required to run contracts: {}", e)
+                            }
+                            None => VMError::FunctionCallError(FunctionCallError::LinkError { msg: err.to_string() })
```

Comment:
This looks sketchy -- what kind of other errors can we get here? Can we change upstream API to avoid downcasts here? If we cant, than I think a commment about when this branch happens might be useful.

Reply:
This branch should be unreachable. The only reason we're dealing with a dynamic error here is due to the way crate graph looks like (`wasmer_vm` contains the `Artifact` trait and `wasmer_engine` contains `InstantiationError` – I should just add an associated type to the trait one of those days...)

Reply:
Than maybe put `unreachable` here or at least `tracing::error`?

Reply:
Actually, probably log_assert would be best?

https://github.com/near/nearcore/blob/bcd5744f307ae4bcbf504b358d22b43da58a58a2/core/o11y/src/lib.rs#L186

Reply:
I don't see how that'd work here, since its non-diverging, unless the idea is to also return an error anyway?

Reply:
Year, `log_assert > tracing::error`, and they both are incomparable with `unreachable`